File: apps/ios/BankCore/Networking/APIClient.swift
import Foundation

struct APIError: Error, LocalizedError, Equatable {
    let status: Int
    let detail: String
    var errorDescription: String? { detail }
}

struct BankCoreAPI {
    var session: URLSession = .shared

    func login(taxId: String, password: String) async throws -> TokenResponse {
        try await post(
            "/auth/login",
            body: ["tax_id": TaxID.digits(taxId), "password": password],
            token: nil
        )
    }

    func register(taxId: String, fullName: String, email: String, password: String) async throws -> UserProfile {
        try await post(
            "/auth/register",
            body: [
                "tax_id": TaxID.digits(taxId),
                "full_name": fullName,
                "email": email,
                "password": password,
            ],
            token: nil
        )
    }

    func me(token: String) async throws -> UserProfile {
        try await get("/auth/me", token: token)
    }

    func directory(taxId: String, token: String) async throws -> DirectoryEntry {
        try await get("/auth/directory/\(TaxID.digits(taxId))", token: token)
    }

    func createOrGetAccount(userId: UUID, token: String) async throws -> Account {
        try await post("/accounts/", body: ["user_id": userId.uuidString], token: token)
    }

    func account(id: UUID, token: String) async throws -> Account {
        try await get("/accounts/\(id.uuidString)", token: token)
    }

    func statement(accountId: UUID, token: String) async throws -> [LedgerTransaction] {
        try await get("/accounts/\(accountId.uuidString)/statement", token: token)
    }

    func pix(
        sourceAccountId: UUID,
        destinationKey: String,
        amountReais: Double,
        description: String,
        token: String
    ) async throws -> LedgerTransaction {
        try await post(
            "/transactions/pix",
            body: [
                "source_account_id": sourceAccountId.uuidString,
                "destination_key": TaxID.digits(destinationKey),
                "amount_reais": amountReais,
                "idempotency_key": Idempotency.pix(),
                "description": description,
            ],
            token: token
        )
    }

    private func get<T: Decodable>(_ path: String, token: String?) async throws -> T {
        try await send(path, method: "GET", body: nil as [String: Any]?, token: token)
    }

    private func post<T: Decodable>(_ path: String, body: [String: Any], token: String?) async throws -> T {
        try await send(path, method: "POST", body: body, token: token)
    }

    private func send<T: Decodable>(
        _ path: String,
        method: String,
        body: [String: Any]?,
        token: String?
    ) async throws -> T {
        if token == nil, needsAuth(path) {
            throw APIError(status: 401, detail: "Token de acesso ausente.")
        }

        var request = URLRequest(url: APIConfig.url(path))
        request.httpMethod = method
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token, !token.isEmpty {
            request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        }

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw APIError(status: 0, detail: "Resposta inválida do servidor.")
        }
        if http.statusCode == 401 {
            throw APIError(status: 401, detail: Self.parseDetail(data) ?? "Token de acesso ausente.")
        }
        guard (200..<300).contains(http.statusCode) else {
            throw APIError(status: http.statusCode, detail: Self.parseDetail(data) ?? "Falha na requisição (\(http.statusCode)).")
        }
        do {
            return try BankJSON.decoder.decode(T.self, from: data)
        } catch {
            throw APIError(status: http.statusCode, detail: "Não foi possível ler a resposta da API.")
        }
    }

    private func needsAuth(_ path: String) -> Bool {
        path.hasPrefix("/accounts")
            || path.hasPrefix("/transactions")
            || path.hasPrefix("/auth/me")
            || path.hasPrefix("/auth/directory")
    }

    static func parseDetail(_ data: Data) -> String? {
        guard let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        if let detail = object["detail"] as? String {
            return detail
        }
        if let items = object["detail"] as? [[String: Any]] {
            let messages = items.compactMap { $0["msg"] as? String }
            if !messages.isEmpty { return messages.joined(separator: " ") }
        }
        return nil
    }
}

enum BankJSON {
    static let decoder: JSONDecoder = {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .custom { decoder in
            let container = try decoder.singleValueContainer()
            let raw = try container.decode(String.self)
            if let date = BankDate.parseAPI(raw) { return date }
            throw DecodingError.dataCorruptedError(
                in: container,
                debugDescription: "Unrecognized date: \(raw)"
            )
        }
        return decoder
    }()
}

File: apps/android/app/src/main/java/br/vortex/bankcore/data/ApiClient.kt
package br.vortex.bankcore.data

import br.vortex.bankcore.BuildConfig
import br.vortex.bankcore.util.TaxId
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import java.util.concurrent.TimeUnit

class BankCoreApi(
    private val baseUrl: String = BuildConfig.API_BASE_URL,
) {
    private val json = Json {
        ignoreUnknownKeys = true
        isLenient = true
    }

    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(20, TimeUnit.SECONDS)
        .readTimeout(30, TimeUnit.SECONDS)
        .apply {
            if (BuildConfig.DEBUG) {
                addInterceptor(
                    HttpLoggingInterceptor().apply {
                        level = HttpLoggingInterceptor.Level.BASIC
                    },
                )
            }
        }
        .build()

    private val media = "application/json; charset=utf-8".toMediaType()

    fun login(taxId: String, password: String): TokenResponse =
        post(
            "/auth/login",
            mapOf("tax_id" to TaxId.digits(taxId), "password" to password),
            token = null,
        )

    fun register(taxId: String, fullName: String, email: String, password: String): UserProfile =
        post(
            "/auth/register",
            mapOf(
                "tax_id" to TaxId.digits(taxId),
                "full_name" to fullName,
                "email" to email,
                "password" to password,
            ),
            token = null,
        )

    fun me(token: String): UserProfile = get("/auth/me", token)

    fun directory(taxId: String, token: String): DirectoryEntry =
        get("/auth/directory/${TaxId.digits(taxId)}", token)

    fun createOrGetAccount(userId: String, token: String): Account =
        post("/accounts/", mapOf("user_id" to userId), token)

    fun account(id: String, token: String): Account = get("/accounts/$id", token)

    fun statement(accountId: String, token: String): List<LedgerTransaction> =
        get("/accounts/$accountId/statement", token)

    fun pix(
        sourceAccountId: String,
        destinationKey: String,
        amountReais: Double,
        description: String,
        token: String,
    ): LedgerTransaction = post(
        "/transactions/pix",
        mapOf(
            "source_account_id" to sourceAccountId,
            "destination_key" to TaxId.digits(destinationKey),
            "amount_reais" to amountReais,
            "idempotency_key" to Idempotency.pix(),
            "description" to description,
        ),
        token,
    )

    private inline fun <reified T> get(path: String, token: String?): T =
        send(path, "GET", null, token)

    private inline fun <reified T> post(path: String, body: Map<String, Any?>, token: String?): T =
        send(path, "POST", body, token)

    private inline fun <reified T> send(
        path: String,
        method: String,
        body: Map<String, Any?>?,
        token: String?,
    ): T {
        if (token.isNullOrBlank() && needsAuth(path)) {
            throw ApiException(401, "Token de acesso ausente.")
        }
        val url = baseUrl.trimEnd('/') + if (path.startsWith("/")) path else "/$path"
        val builder = Request.Builder()
            .url(url)
            .header("Accept", "application/json")
        if (!token.isNullOrBlank()) {
            builder.header("Authorization", "Bearer $token")
        }
        if (body != null) {
            val obj = buildJsonObject {
                body.forEach { (k, v) ->
                    put(
                        k,
                        when (v) {
                            null -> JsonNull
                            is Int -> JsonPrimitive(v)
                            is Long -> JsonPrimitive(v)
                            is Double -> JsonPrimitive(v)
                            is Boolean -> JsonPrimitive(v)
                            else -> JsonPrimitive(v.toString())
                        },
                    )
                }
            }
            builder.header("Content-Type", "application/json")
            builder.method(method, obj.toString().toRequestBody(media))
        } else {
            builder.method(method, null)
        }
        client.newCall(builder.build()).execute().use { response ->
            val bytes = response.body?.string().orEmpty()
            if (response.code == 401) {
                throw ApiException(401, parseDetail(bytes) ?: "Token de acesso ausente.")
            }
            if (response.code !in 200..299) {
                throw ApiException(
                    response.code,
                    parseDetail(bytes) ?: "Falha na requisição (${response.code}).",
                )
            }
            return try {
                json.decodeFromString<T>(bytes)
            } catch (_: Exception) {
                throw ApiException(response.code, "Não foi possível ler a resposta da API.")
            }
        }
    }

    private fun needsAuth(path: String): Boolean =
        path.startsWith("/accounts") ||
            path.startsWith("/transactions") ||
            path.startsWith("/auth/me") ||
            path.startsWith("/auth/directory")

    private fun parseDetail(raw: String): String? {
        return try {
            val obj = json.parseToJsonElement(raw).jsonObject
            val detail = obj["detail"] ?: return null
            runCatching { detail.jsonPrimitive.content }.getOrNull()
                ?: runCatching {
                    detail.jsonArray.mapNotNull { item ->
                        runCatching { item.jsonObject["msg"]?.jsonPrimitive?.content }.getOrNull()
                    }.joinToString(" ").ifBlank { null }
                }.getOrNull()
        } catch (_: Exception) {
            null
        }
    }
}
